### PR TITLE
fix: volumeMount create error

### LINKED ISSUE #207: [Bug]: default cn/dn cache volume error
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```Markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93):
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

When we not set cache volume, it will have volumeMount but not have volume created.

### Expected Behavior

using memory storage

### Steps to Reproduce

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by wanglei4687**
Signed-off-by: wanglei4687 <wanglei4687@gmail.com>

**What type of PR is this?**

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

Fixes #207 #243

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/controllers/cnset/resource.go [hunk 1]
```diff
@@ -77,10 +77,27 @@ func syncPodSpec(cn *v1alpha1.CNSet, sts *kruise.StatefulSet, sp v1alpha1.Shared
...
+	dataVolume := corev1.VolumeMount{
+		Name:      common.DataVolume,
+		MountPath: common.DataPath,
 	}
+
+	configVolume := corev1.VolumeMount{
+		Name:      common.ConfigVolume,
+		ReadOnly:  true,
+		MountPath: common.ConfigPath,
+	}
```
Comment (aylei) [inline pkg/controllers/cnset/resource.go:92]: declaration of configVolume and append configVolume can be merged to the declaration of `volumeMountsList`

**Review by aylei — COMMENTED** [review]
the rest LGTM

**Review by aylei — COMMENTED** [review]
please add a test case to cover this fix

**Comment by wanglei4687** [comment]
@mergify update

File: pkg/controllers/cnset/controller_test.go [hunk 2]
```diff
@@ -129,6 +129,29 @@ func TestCNSetActor_Observe(t *testing.T) {
 	}
 }
 
+func Test_syncPods(t *testing.T) {
```
Comment (aylei) [inline pkg/controllers/cnset/controller_test.go:132]: No case？

File: pkg/controllers/dnset/controller_test.go [hunk 3]
```diff
@@ -72,6 +92,17 @@ func TestDNSetActor_Observe(t *testing.T) {
 				g.Expect(action.String()).To(ContainSubstring("Create"))
 			},
 		},
+		{
+			name:  "create when resource not exist and no cache volume",
+			dnset: tplNoVolume,
+			client: &fake.Client{
+				Client: fake.KubeClientBuilder().WithScheme(s).Build(),
+			},
+			expect: func(g *WithT, action recon.Action[*v1alpha1.DNSet], err error) {
```
Comment (aylei) [inline pkg/controllers/dnset/controller_test.go:101]: should also test the sts spec is correct

File: pkg/controllers/dnset/controller_test.go [hunk 4]
```diff
@@ -151,6 +197,77 @@ func TestDNSetActor_Observe(t *testing.T) {
...
+				},
+				Status: v1alpha1.LogSetStatus{
+					Discovery: &v1alpha1.LogSetDiscovery{
+						Port:    6001,
+						Address: "test",
+					},
+				},
+			}
+			syncPodSpec(tt.dnset, tt.sts, tt.sp)
+			if len(tt.sts.Spec.Template.Spec.Containers[0].VolumeMounts) != 1 {
```
Comment (aylei) [inline pkg/controllers/dnset/controller_test.go:264]: `length == 1` is not always the case, nor can it ensure the sts spec is valid

File: pkg/utils/utils.go [hunk 5]
```diff
@@ -39,3 +39,12 @@ func IsServiceReady(svc *corev1.Service) bool {
 
 	return true
 }
+
+func DiffVolumeMount(key string, list []corev1.VolumeMount) bool {
```
Comment (aylei) [inline pkg/utils/utils.go:43]: what does `DiffVolumeMount` mean?

File: pkg/controllers/dnset/controller_test.go [hunk 6]
```diff
@@ -151,6 +202,86 @@ func TestDNSetActor_Observe(t *testing.T) {
...
+						Address: "test",
+					},
+				},
+			}
+			syncPodSpec(tt.dnset, tt.sts, tt.sp)
+			if tt.dnset.Spec.CacheVolume == nil {
+				// if cacheVolume not set, volumeClaimTemplates should be 0
+				// dataVolumeMount should not be created.
+				if len(tt.sts.Spec.VolumeClaimTemplates) == 0 {
+					if utils.DiffVolumeMount(common.DataVolume, tt.sts.Spec.Template.Spec.Containers[0].VolumeMounts) {
```
Comment (aylei) [inline pkg/controllers/dnset/controller_test.go:273]: other logic may add `VolumeClaimTemplates` and should be relevant to this test case, the thing we want to test is that their is no cache volume in `VolumeClaimTemplates`.

**Review by aylei — APPROVED** [review]
lgtm

**Pull request merged**